stdlib/dynarray.mli

```diff
@@ -0,0 +1,619 @@
+
+val of_seq : 'a Seq.t -> 'a t
+(** [of_seq seq] is an array containing the same elements as [seq].
+
+    It traverses [seq] once and will terminate only if [seq] is finite. *)
+
+val to_seq : 'a t -> 'a Seq.t
+(** [to_seq a] is the sequence of elements
+    [get a 0], [get a 1]... [get a (length a - 1)].
+
+    Because sequences are computed on-demand, we have to assume that
+    the array may be modified in the meantime. We give a precise
+    specification for this case below.
+
+    Demanding the [i]-th element of the resulting sequence (which can
+    happen zero, one or several times) will access the [i]-th element
+    of [a] at the time of the demand. The sequence stops if [a] has
+    less than [i] elements at this point.
+*)
+
+val to_seq_rev : 'a t -> 'a Seq.t
+(** [to_seq_rev a] is the sequence of elements
+    [get a (l - 1)], [get a (l - 2)]... [get a 0],
+    where [l] is [length a] at the time [to_seq_rev] is invoked.
+
+    Elements that have been removed from the array by the time they
+    are demanded in the sequence are skipped.
+*)
+
+
+(** {1:advanced Advanced topics for performance} *)
+
+(** {2:capacity Backing array, capacity}
+
+    Internally, a dynamic array uses a {b backing array} (a fixed-size
+    array as provided by the {!Array} module) whose length is greater
+    or equal to the length of the dynamic array. We call {b capacity}
+    the length of the backing array.
+
+    The capacity of a dynamic array is relevant in advanced scenarios,
+    when reasoning about the performance of dynamic array programs:
+    {ul
+    {- The memory usage of a dynamic array is proportional to its capacity,
+       rather than its length.}
+    {- When then is no empty space left at the end of the backing array.
+       adding elements requires allocating a new, larger backing array.}}
+
+    The implementation uses a standard exponential reallocation
+    strategy which guarantees amortized constant-time operation: the
+    total capacity of all backing arrays allocated over the lifetime
+    of a dynamic array is at worst proportional to the total number of
+    elements added.
+
+    In other words, users need not care about capacity and reallocations,
+    and they will get reasonable behavior by default. However, in some
+    performance-sensitive scenarios the functions below can help control
+    memory usage or guarantee an optimal number of reallocations.
+*)
+
+val capacity : 'a t -> int
+(** [capacity a] is the length of [a]'s backing array. *)
+
+val ensure_capacity : 'a t -> int -> unit
+(** [ensure_capacity a n] makes sure that the capacity of [a]
+    is at least [n].
+
+    @raise Invalid_argument if the requested capacity is
+      outside the range [0 .. Sys.max_array_length].
+
+    An example would be to implement {!of_array} (without using {!init} directly):
+    {[
+    let of_array arr =
+      let a = Dynarray.create () in
+      Dynarray.ensure_capacity a (Array.length arr);
```

Comment:
The natural way to do this would be to add an optional element to all array-creating operations, so `make` and `init` but also `of_{array,list,seq}`, etc. I am not fond of this approach because it causes a lot of conceptual duplication -- a pollution of initial configuration in many functions, that only grows worse as we think of more initial configuration setting (resizing strategy, etc.).

For the advanced use-cases where users want this, I prefer to make it easy to think of how to use `create` followed by `ensure_capacity`, and then populating functions (`add_*`). (It would be nice to have an `add_init` function or some other way to add with a generic `(int -> 'a)` function, but I am not sure what would be a good API for this so I prefer not to try to add something at this relatively late point in the review process.)